File: MajorityElementII.py
"""
229. Majority Element II
Given an integer array of size n, find all elements that appear more than ⌊ n/3 ⌋ times. The algorithm should run in linear time and in O(1) space.
"""
# O(n)
# O(1)
# Moore's Voting
class Solution(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums:
            return []
        lookup = {}
        for i in range(len(nums)):
            if len(lookup) < 2:
                if nums[i] not in lookup:
                    lookup[nums[i]] = 1
                else:
                    lookup[nums[i]] += 1
            else:
                if nums[i] in lookup:
                    lookup[nums[i]] += 1
                else:
                    for key in lookup:
                        if lookup[key] == 0:
                            lookup.pop(key)
                            lookup[nums[i]] = 1
                            break
                    else:
                        for key in lookup:
                            lookup[key] -= 1
        res = []
        for key in lookup:
            count = 0
            for num in nums:
                if num == key:
                    count += 1
            if count > len(nums) // 3:
                res.append(key)
        return res

class Solution2(object):
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        num1, num2, count1, count2 = None, None, 0, 0
        for num in nums:
            if num == num1:
                count1 += 1
            elif num == num2:
                count2 += 1
            elif not count1:
                num1, count1 = num, 1
            elif not count2:
                num2, count2 = num, 1
            else:
                count1, count2 = count1 - 1, count2 - 1
        count1, count2 = 0, 0
        for num in nums:
            if num == num1:
                count1 += 1
            elif num == num2:
                count2 += 1
        res = []
        if count1 > len(nums) // 3:
            res.append(num1)
        if count2 > len(nums) // 3:
            res.append(num2)
        return res

if __name__ == "__main__":
    print(Solution2().majorityElement([1, 1, 1, 2, 3, 2, 4, 6]))

File: Java/MajorityElementII.java
import java.util.*;
public class MajorityElementII {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> res = new ArrayList<Integer>();
        int[] twoNums = new int[2];
        int[] count = new int[2];
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] == twoNums[0]) {
                count[0]++;
            } else if(nums[i] == twoNums[1]) {
                count[1]++;
            } else if(count[0] == 0) {
                twoNums[0] = nums[i];
                count[0] = 1;
            } else if(count[1] == 0) {
                twoNums[1] = nums[i];
                count[1] = 1;
            } else {
                count[0]--;
                count[1]--;
            }
        }
        count[0] = 0;
        count[1] = 0;
        for(int i = 0; i < nums.length; i++) {
            if(nums[i] == twoNums[0]) {
                count[0]++;
            } else if(nums[i] == twoNums[1]) {
                count[1]++;
            }
        }
        if(count[0] > nums.length / 3) {res.add(twoNums[0]);}
        if(count[1] > nums.length / 3) {res.add(twoNums[1]);}
        return res;
    }

}
